File: app/core/events.py
"""通知イベントの中間表現。

**`notifiers/` は「誰に何を送るか」だけを知る。CP のリソースや項目 ID を知らない**
（`rules/50-code-style.md`）。ウォッチャーがこの中間表現を組み立てて渡す。

通知本文は**最初の取得時に組み立てて保存する。**
再送のためにイベントを再取得しない（`rules/20-rate-limit.md`）。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class Notification:
    """1 通の通知。

    Attributes:
        watcher_id:  発生元のウォッチャー
        resource_id: CP のリソース ID。冪等キーの一部
        event_type:  `item_changed` / `status_changed` / `created` など
        digest:      通知本文を決定づける値のハッシュ。冪等キーの一部
        channel_key: 送信先の論理名。Slack のチャンネル、またはメール宛先の解決キー
        subject:     件名（メール用。Slack では本文の先頭に使う）
        body:        本文（組み立て済み）
        to_address:  メールの宛先。Slack 通知では None
        meta:        ログ用の補助情報。**個人情報を入れない**
    """

    watcher_id: str
    resource_id: str
    event_type: str
    digest: str
    channel_key: str
    subject: str
    body: str
    to_address: str | None = None
    meta: dict[str, Any] = field(default_factory=dict)

    def idempotency_key(self) -> tuple[str, str, str, str]:
        return (self.watcher_id, self.resource_id, self.event_type, self.digest)


@dataclass
class CycleResult:
    """1 サイクルの結果。スケジューラがこれを見てカーソルと失敗カウンタを更新する。"""

    ok: bool
    requests_used: int = 0
    events_detected: int = 0
    events_notified: int = 0
    exhausted: bool = False
    error: str | None = None

    def as_log_fields(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "requests_used": self.requests_used,
            "events_detected": self.events_detected,
            "events_notified": self.events_notified,
            "exhausted": self.exhausted,
            "error": self.error,
        }

File: gas/src/core/events.js
/**
 * 1サイクルの結果（Python 版 `app/core/events.py` の `CycleResult`）。
 *
 * **Runner はこの戻り値だけを見てカーソルを進めるかどうかを決める**（仕様書 11.4）。
 * ウォッチャー自身にカーソルを書かせない。コミット点を1箇所に閉じるため。
 *
 * 通知の中間表現（`Notification`）もここに置く。**ウォッチャーが組み立て、
 * notifiers/ が送る。**notifiers/ は CP のリソースも項目 ID も知らない
 * （rules/50-code-style.md）。
 */
const Events = (function () {

  /**
   * @param fields
   *   ok               成功したか
   *   exhausted        予算切れ・時間切れで打ち切ったか（異常ではない）
   *   cursor           コミットするカーソル。`{ value: Date, pageOffset, bootstrapped }`
   *                    **これを入れたサイクルだけカーソルが前進する。**
   *                    入れなければ前進しない（＝次サイクルで再処理される）
   *   eventsDetected   検知した件数
   *   eventsNotified   実際に送った件数
   *   error            失敗した理由（ok=false のとき）
   */
  function result(fields) {
    const f = fields || {};
    return {
      ok: f.ok !== false,
      exhausted: !!f.exhausted,
      cursor: f.cursor || null,
      eventsDetected: f.eventsDetected || 0,
      eventsNotified: f.eventsNotified || 0,
      requestsUsed: f.requestsUsed || 0,
      skipped: f.skipped || null,     // ロック取得失敗・自動停止のときだけ入る
      error: f.error || null,
    };
  }

  /** 予算切れ・時間切れ。**カーソルも snapshots も進めない**（仕様書 11.4）。 */
  function exhausted(fields) {
    const f = fields || {};
    f.ok = true;
    f.exhausted = true;
    f.cursor = null;
    return result(f);
  }

  function failed(error, fields) {
    const f = fields || {};
    f.ok = false;
    f.cursor = null;
    f.error = error;
    return result(f);
  }

  /** そもそも実行しなかった（ロックが取れない・自動停止中）。 */
  function skipped(reason) {
    return result({ ok: true, skipped: reason });
  }

  /**
   * 通知の中間表現。
   *
   * @param fields
   *   watcherId / resourceId / eventType / digest  冪等キーの4つ組（rules/30）
   *   channelKey  論理的な宛先。**Webhook URL ではない**（notifiers が解決する）
   *   subject     件名。Slack では本文に含めないが、メール（要件4）で使う
   *   body        送信する本文。**組み立て済みで渡す。**再取得しない（rules/20）
   *   to          個別の宛先（要件4のメールアドレス）。チャンネルが宛先を決める
   *               Slack では使わない。**空のまま通知を作らないこと**
   *               （要件4は宛先が無ければそもそも送らない。仕様書 3.3.6）
   *   meta        ログ用の補助情報。個人情報を入れないこと。
   *               **`to` はここに入れない**（dead_letter に載って共有範囲が広がる）
   */
  function notification(fields) {
    const f = fields || {};
    if (!f.watcherId || !f.resourceId || !f.eventType || !f.digest) {
      throw Errors.config('notification needs watcherId/resourceId/eventType/digest');
    }
    if (!f.channelKey) throw Errors.config('notification needs channelKey');
    return {
      watcherId: f.watcherId,
      resourceId: String(f.resourceId),
      eventType: f.eventType,
      digest: f.digest,
      channelKey: f.channelKey,
      subject: f.subject || '',
      body: f.body || '',
      to: f.to || null,
      meta: f.meta || {},
    };
  }

  /** 「何件取得して何件通知したか」を1行で出すためのログ項目（rules/50-code-style.md）。 */
  function logFields(r) {
    return {
      ok: r.ok,
      requests_used: r.requestsUsed,
      events_detected: r.eventsDetected,
      events_notified: r.eventsNotified,
      exhausted: r.exhausted,
      committed: !!r.cursor,
      skipped: r.skipped,
      error: r.error,
    };
  }

  return {
    result: result,
    notification: notification,
    exhausted: exhausted,
    failed: failed,
    skipped: skipped,
    logFields: logFields,
  };
})();
